Graphical tweaks and bug fixes

mepland opened the pull request:
* Improved `dtreeviz()` plots:
   * Added `leaf_plot_type` parameter. Default is `pie` which replicates prior plots, but now includes a `barh` option!
   * Added `all_axes_spines` parameter, which turns on axes frames on all 4 sides when selected.
   * Allow different left and right arrow colors via `larrow` and `rarrow`. Current `arrow` parameter still works for both.
   * Padded `root_llabel` and `root_rlabel` strings with starting spaces to move inequalities away from the arrows.
   * Fixed incorrect range calculation in `class_split_viz()`, but set the range to the min/max values instead as that makes the visuals appear cleaner.
   * Used `colors["text"]` instead of hard coded `#444443` in digraph, and included `fontcolor` parameters where missing.
   * Removed hard coded Helvetica `fontname` in digraph.
* Improved `ctreeviz_univar()`:
   * Now takes one feature to plot via the `feature_name` string parameter.
   * Fixed a crash when a histogram bin has 0 counts.
   * Splits are now plotted at a consistent y height. Previously would shrink during the for loop as they were set to the current axis height.
   * Added `ticks_fontsize` and `show` parameters to match other functions.
   * Set histogram edge colors via `rect_edge` instead of `edge` to match other functions.
   * Added `tesselation_alpha` to control alpha.
* Improved `ctreeviz_bivar()`:
   * Fixed `angle` matplotlib warning in `patches.Rectangle()`.
   * Added `ticks_fontsize` parameter to match other functions.
   * Added `tesselation_alpha` to control alpha.
* Improved `ctreeviz_leaf_samples()`:
   * Added `label_all_leafs` parameter which turns on (off) individual ticks for each leaf, useful when there are few (many) leafs.
* Improved axes graphics in `ctreeviz_univar(), ctreeviz_bivar(), viz_leaf_samples(), viz_leaf_criterion(), ctreeviz_leaf_samples(), viz_leaf_target()`.
   * Note, did not update the regression tree versions.
* Defined a set of matplotlib default colors, C0, C1, C2, ..., for convenience. See `mpl_colors`, which mirrors `color_blind_friendly_colors`.
* Do not draw the legend frame when `colors['legend_edge']` is None.
* Fixed python text warning in `utils.py`: `pattern = re.compile(r"scale\([0-9.]+\ [0-9.]+\)")`.
* Updated `ax.grid(b=grid)` to new `ax.grid(visible=grid)` syntax throughout.

Note, tessellation is misspelled as "tesselation" throughout the package, but I am not sure if it is worth a breaking changes to correct.

tlapusan commented:
Thanks @mepland for this PR, it seems to contain a lot of good work :) I'm curious for what are you using dtreeviz if you can share it:) I'll look on it over the next days, we are also pretty busy with a big refactoring.

mepland commented:
@tlapusan sure!

[Here is a quick demo notebook](https://github.com/mepland/dtreeviz/blob/8a8755f23b3e05c0dbf7d62386821849a811a45f/notebooks/demo_dtreeviz.ipynb). pdf versions of plots are in `output_dtreeviz_demo.tar.gz` [here](https://github.com/mepland/dtreeviz/blob/8a8755f23b3e05c0dbf7d62386821849a811a45f/notebooks/output_dtreeviz_demo.tar.gz).

We can definitely remove the demo files before the merge.

tlapusan reviewed (commented):
Thanks for this PR, I added some comments :)

mepland commented on dtreeviz/colors.py:
```diff
@@ -31,6 +31,17 @@
 ]
 
 
+mpl_colors = [
```
I didn't want to overwrite the default `color_blind_friendly_colors` in `'classes': color_blind_friendly_colors`, but wanted to have the default matplotlib colors predefined for the user to swap them in via the colors parameter if they wish:
```
from dtreeviz.colors import mpl_colors
trees.dtreeviz(shadow_tree, colors={'classes': mpl_colors})
```

mepland commented on dtreeviz/trees.py:
```diff
@@ -515,6 +536,7 @@ def dtreeviz(tree_model,
              max_X_features_LR: int = 10,
              max_X_features_TD: int = 20,
              depth_range_to_display: tuple = None,
+             all_axes_spines: bool = False,
```
It's turned off by default, but some fields want plots to have boundaries on all sides. I'd rather just leave it in, but we can take it out if needed.

mepland commented on dtreeviz/trees.py:
```diff
@@ -597,7 +621,7 @@ def split_node(name, node_name, split):
             </table>"""
         else:
-            html = f"""<font face="Helvetica" color="#444443" point-size="12">{name}@{split}</font>"""
+            html = f"""<font face="{fontname}" color="{colors["text"]}" point-size="12">{name}@{split}</font>"""
```
Shouldn't it match the font used in the rest of the plots though?

mepland commented on dtreeviz/trees.py:
```diff
@@ -1038,11 +1075,16 @@ def _get_bins(overall_range, nbins_):
 
-    ax.set_xlim(*overall_feature_range_wide)
+    # ax.set_xlim(*overall_feature_range_wide)
+    ax.set_xlim(*overall_feature_range)
```
Fair enough, I thought it looked better without the gap, but can revert it. Will still keep the corrected calculation of `overall_feature_range_wide`.

mepland replied:
https://github.com/parrt/dtreeviz/pull/200/commits/7f08a087395f7a4a4dae07441338eb6af5cc1c29

mepland commented on dtreeviz/trees.py:
```diff
@@ -1281,11 +1330,15 @@ def draw_legend(shadow_tree, target_name, filename, colors=None, fontname="Arial
         boxes.append(box)
 
+    draw_legend_frame=True
+    if colors['legend_edge'] is None:
```
True, and this could be fixed with `colors.get('legend_edge', None)` or something similar, **but** this is also a problem with **all** of the `colors` keys. For example:
```
    ax.scatter(x_data, y_data, marker='o', alpha=colors['scatter_marker_alpha'], c=colors['scatter_marker'],
               s=markersize,
               edgecolor=colors['scatter_edge'], lw=.3)
```

Do you want to fix all of the potential KeyErrors, maybe by checking the `colors` dict once upfront, or just let KeyErrors occur if the user misses a needed key and let them sort it out?

mepland commented on dtreeviz/trees.py:
```diff
@@ -505,6 +525,7 @@ def dtreeviz(tree_model,
              precision: int = 2,
              orientation: ('TD', 'LR') = "TD",
              instance_orientation: ("TD", "LR") = "LR",
+             leaf_plot_type: ('pie', 'barh') = 'pie',
```
Fair enough, reworked to use the `n=` style of the pie charts:
![Screenshot from 2022-11-12 11-58-02](https://user-images.githubusercontent.com/4729931/201485463-5ea379ce-559a-4478-a51e-2334ae2d2d7e.png)

mepland replied:
https://github.com/parrt/dtreeviz/pull/200/commits/585d7604c77e4563959c7d66bddbcd0c97db106d

mepland commented on dtreeviz/trees.py:
```diff
@@ -305,22 +305,26 @@ def plane(node, bbox, color_spectrum):
 
 
 def ctreeviz_univar(tree_model,
```
I added some code to handle this use case: https://github.com/parrt/dtreeviz/pull/200/commits/2b511ea999f740f7a6b05595a6cb674e979ef1b9

mepland replied:
Note that you can get the lower prediction bars back by adding `'preds'` to `show`.

mepland replied:
BTW the `partitioning.ipynb` notebook is throwing warnings in the latest sklearn version:
```
sklearn/tree/_classes.py:404: FutureWarning: Criterion 'mae' was deprecated in v1.0 and will be removed in version 1.2. Use `criterion='absolute_error'` which is equivalent.
sklearn/base.py:450: UserWarning: X does not have valid feature names, but DecisionTreeRegressor was fitted with feature names
```
and the `figsize = (6,2)` lines are not doing anything.

mepland commented:
@tlapusan I've replied to all of your comments, please let me know how we can resolve the open ones and I'll go from there!

mepland commented:
@tlapusan @parrt Just checking back in, what is the status on merging these changes? Thanks!

mepland commented:
@tlapusan @parrt I also just added an optional `leaf_predictions` parameter that lets you pass a dictionary of `{node_id: value}` to annotate the leaves with the raw values from the tree.

This is useful for things like [XGBoost](https://xgboost.readthedocs.io/en/stable/tutorials/model.html#decision-tree-ensembles) when looking at a single tree, as you may want to see the leaf values that are summed in the ensemble and sent to the logistic function, not just the argmax class label.

I added this functionality via a parameter as I want to support FIGS models from [imodels](https://github.com/csinva/imodels) which can otherwise be loaded with `ShadowSKDTree`. Instead of using a parameter, it could be integrated into the shadow tree classes, e.g. `ShadowXGBDTree` class for XGBoost, but then I'd need to make a whole `ShadowFIGSDTree` class to support my FIGS use case.

tlapusan commented:
hi @mepland, we are also working on a big API change and I have to be focus there. I will have time to look over all your changes only after this release. You did also a lot of changes and we need to look closely on the changes, especially on the viz adjustment parts. 

The 'disadvantage' is that the new release will have a new API and some of your changes would need to be moved to the new API. Sorry for that but we can work together and get it done. Thanks for you efforts of making a better dtreeviz library.

parrt commented:
Hi. Just released 2.0.0: https://github.com/parrt/dtreeviz/releases/tag/2.0.0 so you can check out the new API and code reorg.

We should discuss your various goals as enumerated above so we can make decisions about which improvements to grab :)

mepland commented:
Hi @parrt and @tlapusan congrats on version 2.0! Happy to discuss and refactor as needed. Would you like to talk via comments here or have a live chat? Either works for me, I am off this week so the timing is good.

parrt commented:
I think @tlapusan is off mostly but maybe we can take the items in small logical chunks, one per new PR?  Much easier to evaluate :)  I'm free this week

mepland commented:
@parrt Sure! I will port over my tweaks individually in new branches over the coming days and weeks. I ended up tweaking a lot more than I originally planned, hence the very large PR. It will be good to separate the changes out.

mepland commented:
Hi @parrt and @tlapusan just wanted to check in!

I have most of the changes from this massive PR ported over now, between the merged PRs and these 4 open PRs:
- https://github.com/parrt/dtreeviz/pull/220
- https://github.com/parrt/dtreeviz/pull/224
- https://github.com/parrt/dtreeviz/pull/226
- https://github.com/parrt/dtreeviz/pull/227

I have 4 more action items I would like to address. Please let me know your thoughts:
1. Re-implement my leaf_predictions parameter from [here](https://github.com/parrt/dtreeviz/pull/200#issuecomment-1329986770).
@parrt let me know if you would rather do this via the shadow classes, or some other way, but I think the parameter is easiest and is still a flexible solution.

3. Improve axes graphics formatting with util function similar to `add_classifier_legend()`.
We can cleanup a lot of commands like these and make everything more standardized:
```
ax.spines['top'].set_visible(False)
ax.spines['right'].set_visible(False)
ax.spines['left'].set_linewidth(.3)
ax.spines['bottom'].set_linewidth(.3)
ax.set_xlabel(xlabel, fontsize=fontsize, fontname=fontname, color=colors['axis_label'])
ax.set_ylabel(ylabel, fontsize=fontsize, fontname=fontname, color=colors['axis_label'])
ax.grid(visible=grid)
ax.tick_params(axis='both', which='major', width=.3, labelcolor=colors['tick_label'], labelsize=ticks_fontsize)
or
ax.tick_params(colors=colors['tick_label'])
```

Speaking of `add_classifier_legend()`, would it be a good idea to use `add_classifier_legend()` inside `_draw_legend()`? Much of the central code is the same, it would just need a few commands pre/post `add_classifier_legend()` call?

4. Let the user specify what feature, or 2 features, by name to plot in `ctree_feature_space()`, rather than require that the `shadow_tree` itself only have 1 or 2 features.
Even if the tree has multiple features, it is still nice to be able to quickly plot a single feature and all of the splits on that feature - without resorting to an entirely different tree trained on just that feature(s). @parrt let me know if you'd like this created as an issue, as it is a larger enhancement.

5. `tessellation` is misspelled as "`tesselation`" throughout the package.
As release 2.0 already introduced breaking changes, it may be a good time to fix this typo for 2.1.

Additionally I just created https://github.com/parrt/dtreeviz/issues/228 as an enhancement, and noticed a bug in https://github.com/parrt/dtreeviz/issues/225.

parrt commented:
> 5\. `tessellation` is misspelled as "`tesselation`" throughout the package.

Added https://github.com/parrt/dtreeviz/pull/230

parrt commented:
>  Improve axes graphics formatting with util function similar to add_classifier_legend(). We can cleanup a lot of commands like these and make everything more standardized:

If you're willing to do the refactoring in a PR to see how it looks, happy to check it out.

mepland commented:
> > Improve axes graphics formatting with util function similar to add_classifier_legend(). We can cleanup a lot of commands like these and make everything more standardized:
> 
> If you're willing to do the refactoring in a PR to see how it looks, happy to check it out.

See https://github.com/parrt/dtreeviz/pull/232

parrt commented:
@tlapusan @mepland hiya. I need to cut a 2.1 release so I can start working on a tutorial. I'd like to do this Friday California time. What is the minimal set of things we need to go into 2.1? Of course we can make changes after this for 2.x :)

mepland commented:
@parrt Before 2.1 I'd like to:

- [ ] Merge https://github.com/parrt/dtreeviz/pull/232
- This PR is ready to go, just needs a review and merge.
- [ ] [Remove the tesselation_alpha parameter in `_ctreeviz_univar()`.](https://github.com/parrt/dtreeviz/pull/227#issuecomment-1372839121)
- Will do in a new branch & PR tonight.
- [ ] Re-implement my leaf_predictions parameter from [here](https://github.com/parrt/dtreeviz/pull/200#issuecomment-1329986770)
- @parrt let me know if you would rather do this via the shadow classes, or some other way, but I think the parameter is easiest and is still a flexible solution. If you are okay with using the parameter I can re-add the code tonight. If you want to build the functionality into the shadow classes let's make an issue and push it back to a later release.
- [ ] Hide wedge for classifiers if wedge near edge: https://github.com/parrt/dtreeviz/issues/113
- I can do this tonight if @parrt can answer my questions [here](https://github.com/parrt/dtreeviz/issues/113#issuecomment-1371533256).

mepland commented:
> > `leaf_predictions` parameter
> 
> I don't see how the tree is different there and how would we show different values without screwing up the plot possibly?

The leafs are labeled with the actual values, rather than the classes as chosen by an `argmax`. Note that one is even negative (-0.00) here, this is XGboost so the leaf values (XGBoost calls them weights in it's documentation) are summed together and then a sigmoid is taken.

It would just be nice to have the option between plotting the predicted class, as is currently done by `argmax`, or showing the value returned by the leaf.

![xgboost_demo](https://user-images.githubusercontent.com/4729931/204423501-19b236f0-bec1-461f-af38-b7b858461731.png)

parrt commented:
Oh that is interesting. Well we definitely want to see the prediction there no matter what. I don't think we need an option. Somehow this is different than for other supported decision tree libraries I guess?

parrt commented:
By the way when I look at the Colors notebook I see some improper rendering:

<img width="187" alt="Screen Shot 2023-01-05 at 4 02 39 PM" src="https://user-images.githubusercontent.com/178777/210903297-10e36bf7-5a29-4c44-b055-8457eacacadb.png">

Is this something that we inadvertently messed up when we changed the padding around edge labels? Oh actually that is in the master branch so maybe it's better in the dev branch. I'm on the wrong machine at the moment at work so I can't check it

tlapusan commented:
> By the way when I look at the Colors notebook I see some improper rendering:
> 
> <img alt="Screen Shot 2023-01-05 at 4 02 39 PM" width="187" src="https://user-images.githubusercontent.com/178777/210903297-10e36bf7-5a29-4c44-b055-8457eacacadb.png">
> 
> Is this something that we inadvertently messed up when we changed the padding around edge labels? Oh actually that is in the master branch so maybe it's better in the dev branch. I'm on the wrong machine at the moment at work so I can't check it

looks fine for me also in dev branch.

tlapusan commented:
> @tlapusan @mepland hiya. I need to cut a 2.1 release so I can start working on a tutorial. I'd like to do this Friday California time. What is the minimal set of things we need to go into 2.1? Of course we can make changes after this for 2.x :)

It's fine for me to make a new release with what we already have merged. I will have time in this weekend to look over what you discussed, PRs, etc.

tlapusan commented:
> The leafs are labeled with the actual values

do you have a viz example ? I saw that you already implemented this is the initial PR. Would help to better understand.
From your code I understand that you want to show these values into the graph label, 
`label=f"n={nsamples}\n{prediction}`

if these if the case, we will have situations when a leaf will have plenty of values and would be challenging to display them.

parrt commented:
> Would you like me to create an issue to track this?

Sure. We can make this a feature at some point.

mepland commented:
Done, see https://github.com/parrt/dtreeviz/issues/239

mepland commented:
> if these if the case, we will have situations when a leaf will have plenty of values and would be challenging to display them.

@tlapusan do you mean for multiclass models?

tlapusan commented:
> > if these if the case, we will have situations when a leaf will have plenty of values and would be challenging to display them.
> 
> @tlapusan do you mean multiclass models?

from your initial PR (#200) code, I understood that instead of the argmax class label you wanted to put a list of weigths as a label. Maybe I got it wrong, that's why I said that something visual will help to better understand what you suggested.

mepland commented:
> from your initial PR (#200) code, I understood that instead of the argmax class label you wanted to put a list of weigths as a label. Maybe I got it wrong, that's why I said that something visual will help to better understand what you suggested.

@tlapusan no, in my original solution I just provided a dict of `{node_id: weight}` to label the leaves. The output is pretty much the same, just with one weight printed below `n=` for each leaf, instead of the predicted class name. A graphical example is:

> ![xgboost_demo](https://user-images.githubusercontent.com/4729931/204423501-19b236f0-bec1-461f-af38-b7b858461731.png)

parrt commented:
So this would be an option that instead of showing "survive" or "perish", it would show the probability of the majority class? Or I guess xgboost is calling them weights or something? So I guess it could be different values depending on the library?

mepland commented:
> So this would be an option that instead of showing "survive" or "perish", it would show the probability of the majority class? Or I guess xgboost is calling them weights or something? So I guess it could be different values depending on the library?

@parrt exactly!

parrt commented:
ok, is there a name for the value from each lib?

mepland commented:
> ok, is there a name for the value from each lib?

@parrt

For `XGBoost` it is the `Gain`:
```
dfp_xgboost = model.get_booster().trees_to_dataframe()
dfp_xgboost_0 = dfp_xgboost.loc[( (dfp_xgboost['Tree'] == 0) & (dfp_xgboost['Feature'] == 'Leaf') )]
leaf_predictions_xgboost_0 = {}
for _ in dfp_xgboost_0.to_dict('records'):
    leaf_predictions_xgboost_0[_['Node']] = _['Gain']
```


For `sklearn`'s `DecisionTreeClassifier` it is kind of `model.tree_.value`, but that is an array of counts per class seen at the leaf in training. You'd need need to convert to the [fraction of the positive class](https://scikit-learn.org/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html#sklearn.tree.DecisionTreeClassifier.predict_proba) as is done [here](https://github.com/scikit-learn/scikit-learn/blob/98cf537f5/sklearn/tree/_classes.py#L897-L943). Other references [here](https://stackoverflow.com/questions/47251594/scikit-learn-decision-tree-probability-of-prediction-being-a-or-b) and [here](https://stackoverflow.com/questions/47719001/what-does-scikit-learn-decisiontreeclassifier-tree-value-do/47719621#47719621).


I am also contributing to the FIGS ensemble tree model in the `imodels` [package](https://github.com/csinva/imodels). For FIGS the leaf values are residuals fit from the other trees. I wrote a [function](https://github.com/csinva/imodels/blob/c18078ecb919597bdaaae0ca4d521093274f02f2/imodels/tree/viz_utils.py#L60) to convert the FIGS model to a `sklearn` `DecisionTreeClassifier` or `DecisionTreeRegressor`, so we can easily use the `ShadowSKDTree` of `dtreeviz`, and extract the leaf values as a dictionary [here](https://github.com/csinva/imodels/blob/master/imodels/tree/viz_utils.py#L137-L147). I would probably either need to write my own `ShadowFIGSTree` class to load FIGS naively, or hack the `sklearn` `values` array somehow. That's why I just went with passing a dict of `{node_id: value}` initially :smile:

parrt commented:
So is this something that we need to create a new function for that returns in appropriate floating point value or whatever four leaf instead of the predicted class? (just for classifiers)

mepland commented:
> So is this something that we need to create a new function for that returns in appropriate floating point value or whatever four leaf instead of the predicted class? (just for classifiers)

Yep! Or we can leave it the way I originally had it where the user needs to give a dictionary of `{node_id: value_to_print}` and leave the implementation up to them. Probably nicer to build it in as a supported option in each Shadow Class implementation though.

parrt commented:
Seems like people would want always to show the value rather than the prediction versus just for one node, right?

mepland commented:
> Seems like people would want always to show the value rather than the prediction versus just for one node, right?

Well, the dictionary would have all of the leaf `node_id`'s ideally: `node_labels={11: 0.3, 12: 0.9, 15:0.1, ...}`

parrt commented:
That sounds like a hassle. nobody would know ids ;)  Better to be all or none I'd say.

mepland commented:
Closing this PR, as all of its changes have now either been implemented or tracked in their own issues. I'll review later to make sure I didn't forget anything.
